storage/range_tree.go

```diff
@@ -18,31 +18,306 @@
 package storage
 
 import (
+	"github.com/cockroachdb/cockroach/client"
 	"github.com/cockroachdb/cockroach/proto"
 	"github.com/cockroachdb/cockroach/storage/engine"
+	"github.com/cockroachdb/cockroach/util"
 )
 
-// SetupRangeTree creates a new range tree.  This should only be called as part of
-// store.BootstrapRange.
-func SetupRangeTree(batch engine.Engine, ms *engine.MVCCStats,
-	timestamp proto.Timestamp, startKey proto.Key) error {
+type (
+	// RangeTree represents holds the metadata about the RangeTree.  There is
+	// only one for the whole cluster.
+	rangeTree proto.RangeTree
+	// RangeTreeNode represents a node in the RangeTree, each range
+	// has only a single node.
+	rangeTreeNode proto.RangeTreeNode
+)
+
+// TODO(bram): Add a cache
+// TODO(bram): Add parent keys
+// TODO(bram): Add delete for merging ranges.
+// TODO(bram): Add more elaborate testing.
+
+// set saves the RangeTree node to the db.
+func (n *rangeTreeNode) set(db *client.KV) error {
+	if err := db.PutProto(engine.RangeTreeNodeKey(n.Key), (*proto.RangeTreeNode)(n)); err != nil {
```

Comment:
We're really going to want these to be PreparePutProto so all the writes can be done in parallel.